smac/facade/hydra_facade.py

```diff
@@ -130,20 +252,124 @@ def optimize(self):
 
             # modify TAE such that it return oracle performance
             self.tae = ExecuteTARunHydra(ta=self.scenario.ta, run_obj=self.scenario.run_obj,
-                                         cost_oracle=cost_per_inst, tae=self.tae_type)
+                                         cost_oracle=self.cost_per_inst, tae=self._tae)
 
             self.scenario.output_dir = os.path.join(self.top_dir, "smac3-output_%s" % (
                 datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d_%H:%M:%S_%f')))
             self.output_dir = create_output_directory(self.scenario, run_id=self.run_id, logger=self.logger)
-            if i != self.n_iterations - 1:
-                self.solver = SMAC(scenario=self.scenario, tae_runner=self.tae, **self.kwargs)
         self.rh.save_json(fn=os.path.join(self.top_dir, 'all_runs_runhistory.json'), save_external=True)
         with open(os.path.join(self.top_dir, 'portfolio.pkl'), 'wb') as fh:
-            pickle.dump(portfolio, fh)
+            pickle.dump(self.portfolio, fh)
         self.logger.info("~"*120)
         self.logger.info('Resulting Portfolio:')
-        for configuration in portfolio:
+        for configuration in self.portfolio:
             self.logger.info(str(configuration))
         self.logger.info("~"*120)
 
-        return portfolio
+        return self.portfolio
+
+    def _update_portfolio(self, incs: np.ndarray) -> typing.Union[np.float, float]:
+        """
+        Validates all configurations (in incs) and determines which ones to add to the portfolio
+
+        Parameters
+        ----------
+        incs: np.ndarray
+            List of Configurations
+
+        """
+        self.logger.info('*'*120)
+        self.logger.info('Validating')
+        new_rh = self.solver.validate(config_mode=incs,
+                                      instance_mode=self.val_set,
+                                      repetitions=1,
+                                      use_epm=False,
+                                      n_jobs=self.n_optimizers)
+        self.rh.update(new_rh, origin=DataOrigin.EXTERNAL_SAME_INSTANCES)
+        self.logger.info("Number of validated runs: %d", (len(new_rh.data)))
+        self.logger.info('*'*120)
+        self.logger.info('Determining best incumbents')
+        results = []
+        config_cost_per_inst = {}
+        for incumbent in incs:
+            cost_per_inst = new_rh.get_instance_costs_for_config(config=incumbent)
+            config_cost_per_inst[incumbent] = cost_per_inst
+            results.append(np.mean(list(self.cost_per_inst.values())))
+        to_keep_ids = list(map(lambda x: x[0],
+                               sorted(enumerate(results), key=lambda y: y[1])))[:self.incs_per_round]
```

Comment:
Should `incs_per_round` be a fixed parameter? I would iteratively all configurations as long the VBS score improves.

Reply:
Yes I'd say so.
`incs_per_round` gives an upper bound on the number of performed iterations.
If a user were to use Hydra on a scenario where each individual SMAC run takes a day, they could set a limit of `incs_per_round=6` to spend at most 6 days.
So the default value for `incs_per_round` should be rather large but fixed over the hydra call